1 - Add Trending to report

DJensen94 opened the pull request:
Fixes #188 
Add Trending to the credentials and dark web pages of the report
## 🗣 Description ##
This PR adds Trending to two pages in the report to show how many cred breaches/mentions occured compared to the previous weeks
## 💭 Motivation and context ##
We want to give our stakeholders context for the current weeks numbers, seeing how these results compare to past weeks can help them see fluctuations more easily

## ✅ Pre-approval checklist ##

- [X] This PR has an informative and human-readable title.
- [X] Changes are limited to a single goal - *eschew scope creep!*
- [x] All new and existing tests pass.

DJensen94 committed:
Add Trending to report

Add Trending to the credentials and dark web pages of the report

DJensen94 committed:
Remove variable being assigned to itself

remove auto assignation of a variable

aloftus23 committed:
Add test_get_orgs test

aloftus23 reviewed (approved):
Looks good to me.

aloftus23 commented:
@dav3r The PRs are starting to stack up so we are numbering them by priority in the title. This is number 1 when you get the chance to review.

cduhn17 reviewed (approved):
Looks good.

cduhn17 commented:
@dav3r, 
Do you have a time frame of when you might get a chance to take a look at this PR? Thanks.

dav3r commented:
> @dav3r, Do you have a time frame of when you might get a chance to take a look at this PR? Thanks.

I'm going to take a look at it now.

dav3r commented on src/pe_reports/charts.py:
```diff
@@ -88,11 +88,12 @@ def stacked_bar(self):
         plt.rc("axes", axisbelow=True)
         plt.grid(axis="y", zorder=0)
-        plt.xticks(rotation=30, ha="right")
+        plt.xticks(rotation=0)
```
In general, it would be nicer to see smaller, more targeted commits.  That way changes, like all of those to `charts.py` can be documented in the commit log, rather than being hidden amongst a larger commit like https://github.com/cisagov/pe-reports/pull/192/commits/311223a8034efe87cca37fe8e8183daa522b0918.

Some more comments sprinkled in with the code would also greatly improve its readability.

DJensen94 replied:
@dav3r  resolved in commit [453382d](https://github.com/cisagov/pe-reports/pull/192/commits/453382d3e42200b666790f6dd07cf010b8a58c27)

dav3r commented on src/pe_reports/charts.py:
```diff
@@ -152,31 +153,60 @@ def line_chart(self):
+            plt.plot(df.index, df[df.columns[1]], color=color[1], label=df.columns[1])
+
+        plt.ylim(ymin=0, ymax=int(df[df.columns].max().max() * 1.15))
+        # plt.legend(loc=9, ncol=2, framealpha=0, fontsize=8, bbox_to_anchor=(0.5, -0.5))
```
Generally, we recommend removing commented-out/unused code.  If there is a valid reason to keep it around, please add a comment explaining why.

DJensen94 replied:
@dav3r issue resolved in commit [453382d](https://github.com/cisagov/pe-reports/pull/192/commits/453382d3e42200b666790f6dd07cf010b8a58c27)

dav3r commented on src/pe_reports/charts.py:
```diff
@@ -152,31 +153,60 @@ def line_chart(self):
-            )
+        # totals = df.sum(axis=1)
+        # for i, total in enumerate(totals):
+        #     ax.text(totals.index[i], total + 0, round(total), ha="center")
```
See comment above about removing unused code.

DJensen94 replied:
@dav3r issue was resolved in [453382d](https://github.com/cisagov/pe-reports/pull/192/commits/453382d3e42200b666790f6dd07cf010b8a58c27)

dav3r commented on src/pe_reports/charts.py:
```diff
@@ -152,31 +153,60 @@ def line_chart(self):
+                        ),  # distance from text to points (x,y)
+                        ha="center",  # horizontal alignment can be left, right or center
+                        # fontsize=2,
+                    )
```
There is a bunch of duplicated code here.  Are the only values for `len(df.columns)` 1 or 2?  If so, you could add another outer loop like this:
```suggestion
        for col in range(len(df.columns)):
            for i, j in df[df.columns[col]].items():
                if int(j):
                    plt.annotate(
                        str(int(j)),
                        xy=(i, j),
                        textcoords="offset points",  # how to position the text
                        xytext=(
                            0,
                            5,
                        ),  # distance from text to points (x,y)
                        ha="center",  # horizontal alignment can be left, right or center
                    )
```

Note that I removed the commented-out `fontsize=2` line above.

dav3r commented on src/pe_reports/data/db_query.py:
```diff
@@ -61,11 +61,11 @@ def get_orgs(conn):
-def query_hibp_view(org_uid, start_date, end_date):
-    """Query 'Have I Been Pwned?' table."""
+def query_creds_view(org_uid, start_date, end_date):
+    """Query credentials view ."""
```
Remove errant extra space:
```suggestion
    """Query credentials view."""
```

dav3r commented on src/pe_reports/metrics.py:
```diff
@@ -154,24 +130,17 @@ def breach_details(self):
-        )
-        return pw_creds_csg + pw_creds_hibp
+        pw_creds = len(self.creds_view[self.creds_view["password_included"]])
+        return pw_creds
```
Might as well just consolidate these lines:
```suggestion
        return len(self.creds_view[self.creds_view["password_included"]])
```

dav3r commented on src/pe_reports/metrics.py:
```diff
@@ -154,24 +130,17 @@ def breach_details(self):
-        df_cred_hibp = self.query_hibp_view.shape[0]
-        return df_cred_csg + df_cred_hibp
+        df_cred = self.creds_view.shape[0]
+        return df_cred
```
Same here:
```suggestion
        return self.creds_view.shape[0]
```

dav3r commented on src/pe_reports/pages.py:
```diff
@@ -245,16 +247,31 @@ def dark_web(chevron_dict, start_date, end_date, org_uid):
     else:
-        start_date = datetime(end_date.year, end_date.month, 16)
+        start_date = datetime.datetime(end_date.year, end_date.month, 16)
+    days = datetime.timedelta(27)
```
You should add a comment explaining why you are using this 27 day magic number.

DJensen94 replied:
@dav3r resolved in commit [e0a3626](https://github.com/cisagov/pe-reports/pull/192/commits/e0a3626a892cd8b547b59f9ae0aef5a13e503765)

dav3r commented on tests/test_pe_reports.py:
```diff
@@ -104,3 +105,12 @@ def test_reports_bad_log_level():
+    with patch("psycopg2.connect") as mock_connect:
+        mock_connect.cursor.return_value.fetchall.return_value = query_result
+        result = pe_reports.data.db_query.get_orgs(mock_connect)
+    assert result == [("12345", "Test Org 1", "TO_ID_1")]
```
First of all, I applaud you for adding a new test here! 👍   However, I'm not so sure that you are really testing your `get_orgs` query.  If I'm understanding this correctly (which I may not be), you are mocking up the DB connection and telling it what you want the results to be (`query_result`), then you are validating that the query results are the same as what you said you wanted the results to be- which they always should be.  Am I missing something or is the `get_orgs` query not really being tested here?

DJensen94 replied:
@dav3r removed this test in commit [b6288ae](https://github.com/cisagov/pe-reports/pull/192/commits/b6288aeebe7dcb344bb564973970e586e1988fa3)

aloftus23 replied:
Decided to create a new issue to write more useful tests: #198

dav3r replied:
Thanks @DJensen94 and @aloftus23 👍

dav3r reviewed (changes requested):
Take a look at my feedback and please let me know if you have any questions.

DJensen94 committed:
Implement loop for adding data labels

move duplicated code into a loop for adding annotations/data labels to the chart

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Add comments and remove commented out code

Reordered some of the code to fit better in the different comments.

Also removed unnecessary commented out code

DJensen94 committed:
Remove errant extra space

remove space before punctuation

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Remove unnecessary variable assignment

consolidate variables assignement to return statement

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Consolidate variable assignment to the return statement

Move the assignment of variable to the return statement

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Add comment to explain 27 day magic number

Explain why we are using 27 days back for our trending period in a comment

DJensen94 committed:
Merge branch 'DJ_add_trending' of https://github.com/cisagov/pe-reports into DJ_add_trending

DJensen94 committed:
remove unnecessary test

Remove  test get org from db test that doesn't check anything

DJensen94 committed:
Fix return issue for init function

Init function return was not matching the assignment for change to credentials query. This merges the two into one

dav3r commented on src/pe_reports/charts.py:
```diff
@@ -152,32 +161,44 @@ def line_chart(self):
+                        ),  # distance from text to points (x,y)
+                        ha="center",  # horizontal alignment can be left, right or center
+                    )
+        # Specify axis atributes
```
Typo:
```suggestion
        # Specify axis attributes
```

dav3r reviewed (approved):
Once the typo mentioned [here](https://github.com/cisagov/pe-reports/pull/192#discussion_r868409603) is corrected, this one looks good to me. 👍

DJensen94 committed:
Fix comment typo

Fix typo in charts.py

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k reviewed (changes requested):
I'd like to see [the two warnings from LGTM](https://github.com/cisagov/pe-reports/pull/192#issuecomment-1121421689) resolved before merging.

DJensen94 commented:
> I'd like to see [the two warnings from LGTM](https://github.com/cisagov/pe-reports/pull/192#issuecomment-1121421689) resolved before merging.

@dav3r , @jsf9k We are pretty positive that the warnings are unfounded, they claim that origin is not a parameter for 
grouper, but it is a new parameter:
<img width="662" alt="image" src="https://user-images.githubusercontent.com/79864006/167652108-e356ddb9-978f-4513-b898-851240b2f08e.png">
So LGTM is out of date

jsf9k commented:
> > I'd like to see [the two warnings from LGTM](https://github.com/cisagov/pe-reports/pull/192#issuecomment-1121421689) resolved before merging.
> 
> @dav3r , @jsf9k We are pretty positive that the warnings are unfounded, they claim that origin is not a parameter for grouper, but it is a new parameter: <img alt="image" width="662" src="https://user-images.githubusercontent.com/79864006/167652108-e356ddb9-978f-4513-b898-851240b2f08e.png"> So LGTM is out of date

Then you should add [a suppression comment](https://lgtm.com/help/lgtm/alert-suppression#python) for those two warnings.

DJensen94 committed:
add suppression comments to metrics

LGTM is creating warnings that are not relevant, we are adding suppression comments to remove them

DJensen94 committed:
Move suppression comment

Move the suppression comment to the correct line

DJensen94 commented:
> > > I'd like to see [the two warnings from LGTM](https://github.com/cisagov/pe-reports/pull/192#issuecomment-1121421689) resolved before merging.
> > 
> > 
> > @dav3r , @jsf9k We are pretty positive that the warnings are unfounded, they claim that origin is not a parameter for grouper, but it is a new parameter: <img alt="image" width="662" src="https://user-images.githubusercontent.com/79864006/167652108-e356ddb9-978f-4513-b898-851240b2f08e.png"> So LGTM is out of date
> 
> Then you should add [a suppression comment](https://lgtm.com/help/lgtm/alert-suppression#python) for those two warnings.

@jsf9k we have added the suppression comment. LGTM says the warning won't be removed till after the merge occurs.

jsf9k reviewed (approved):
I can't get the LGTM thing to go away either, so I'll have to let it pass for now.
